File: php/x402-gateway/src/Pass.php
<?php
declare(strict_types=1);

namespace Profullstack\X402Gateway;

/** Passes: signed, self-describing tokens, cp_<payload>.<hmac-sha256>. No table behind them. */
final class Pass
{
    public static function b64url(string $bytes): string
    {
        return rtrim(strtr(base64_encode($bytes), '+/', '-_'), '=');
    }

    private static function sign(string $secret, string $data): string
    {
        return self::b64url(hash_hmac('sha256', $data, $secret, true));
    }

    /** @return array{token: string, expires_at: int, ref: ?string} */
    public static function mint(string $secret, ?string $ref, int $expiresAt, ?int $now = null): array
    {
        $now ??= time();
        if ($secret === '') {
            throw new \InvalidArgumentException('a pass needs a signing secret');
        }
        if ($expiresAt <= $now) {
            throw new \InvalidArgumentException('a pass needs a future expiry');
        }
        $payload = self::b64url(json_encode(['v' => 1, 'iat' => $now, 'exp' => $expiresAt, 'ref' => $ref], JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE));
        return ['token' => "cp_$payload." . self::sign($secret, $payload), 'expires_at' => $expiresAt, 'ref' => $ref];
    }

    /** The claims when the signature holds and the pass is live, else null. Never throws on garbage. */
    public static function read(mixed $token, string $secret, ?int $now = null): ?array
    {
        $now ??= time();
        if ($secret === '' || !is_string($token) || !str_starts_with($token, 'cp_')) {
            return null;
        }
        $dot = strpos($token, '.');
        if ($dot === false) {
            return null;
        }
        $payload = substr($token, 3, $dot - 3);
        $sig = substr($token, $dot + 1);
        if ($payload === '' || $sig === '') {
            return null;
        }
        if (!hash_equals(self::sign($secret, $payload), $sig)) {
            return null;
        }
        $raw = base64_decode(strtr($payload, '-_', '+/') . str_repeat('=', (4 - strlen($payload) % 4) % 4), true);
        if ($raw === false) {
            return null;
        }
        $claims = json_decode($raw, true);
        if (!is_array($claims) || ($claims['v'] ?? null) !== 1) {
            return null;
        }
        $exp = $claims['exp'] ?? null;
        if (!is_int($exp) && !is_float($exp)) {
            return null;
        }
        if (is_float($exp) && !is_finite($exp)) {
            return null;
        }
        if ($exp <= $now) {
            return null;
        }
        return ['exp' => $exp, 'iat' => $claims['iat'] ?? null, 'ref' => $claims['ref'] ?? null];
    }
}

File: go/x402gateway/pass.go
package x402gateway

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"math"
	"strings"
	"time"
)

// Pass is a signed, self-describing token: cp_<payload>.<signature>, HMAC-SHA256.
type Pass struct {
	Token     string
	ExpiresAt int64
	Ref       string
}

// Claims are what a valid pass carries.
type Claims struct {
	Exp float64 `json:"exp"`
	Iat any     `json:"iat"`
	Ref any     `json:"ref"`
}

func sign(secret, data string) string {
	m := hmac.New(sha256.New, []byte(secret))
	m.Write([]byte(data))
	return base64.RawURLEncoding.EncodeToString(m.Sum(nil))
}

// MintPass mints a pass. ref may be empty (written as null); now is unix seconds, 0 means the clock.
func MintPass(secret, ref string, expiresAt, now int64) (Pass, error) {
	if now == 0 {
		now = time.Now().Unix()
	}
	if secret == "" {
		return Pass{}, errors.New("a pass needs a signing secret")
	}
	if expiresAt <= now {
		return Pass{}, errors.New("a pass needs a future expiry")
	}
	var refJSON any
	if ref != "" {
		refJSON = ref
	}
	claims := struct {
		V   int   `json:"v"`
		Iat int64 `json:"iat"`
		Exp int64 `json:"exp"`
		Ref any   `json:"ref"`
	}{1, now, expiresAt, refJSON}
	raw, _ := json.Marshal(claims)
	payload := base64.RawURLEncoding.EncodeToString(raw)
	return Pass{Token: "cp_" + payload + "." + sign(secret, payload), ExpiresAt: expiresAt, Ref: ref}, nil
}

// ReadPass returns the claims when the signature holds and the pass is live, else nil. Never panics.
func ReadPass(token, secret string, now int64) *Claims {
	if now == 0 {
		now = time.Now().Unix()
	}
	if secret == "" || !strings.HasPrefix(token, "cp_") {
		return nil
	}
	dot := strings.IndexByte(token, '.')
	if dot < 0 {
		return nil
	}
	payload, sig := token[3:dot], token[dot+1:]
	if payload == "" || sig == "" {
		return nil
	}
	if !hmac.Equal([]byte(sign(secret, payload)), []byte(sig)) {
		return nil
	}
	raw, err := base64.RawURLEncoding.DecodeString(strings.TrimRight(payload, "="))
	if err != nil {
		return nil
	}
	var m map[string]any
	if json.Unmarshal(raw, &m) != nil || m == nil {
		return nil
	}
	if v, ok := m["v"].(float64); !ok || v != 1 {
		return nil
	}
	exp, ok := m["exp"].(float64)
	if !ok || math.IsNaN(exp) || math.IsInf(exp, 0) || exp <= float64(now) {
		return nil
	}
	return &Claims{Exp: exp, Iat: m["iat"], Ref: m["ref"]}
}
